File: framework/runner.py
from __future__ import annotations

from pathlib import Path
import copy
import time
from typing import Dict, List, Tuple

from .adapters import PlatformAdapter
from .errors import make_error
from .models import ErrorCode, RunResult, Status, StepResult
from .utils import make_id, signature_for, utc_now_iso


def _run_step(adapter: PlatformAdapter, command: str, args: Dict) -> Dict:
    if command == "tap":
        return adapter.tap(args)
    if command == "navigate":
        return adapter.navigate(args)
    if command == "act":
        return adapter.act(args)
    if command == "screenshot":
        return adapter.screenshot(args)
    if command == "wait":
        return adapter.wait(args)
    if command == "source":
        return adapter.source(args)
    raise ValueError(f"Unsupported command: {command}")


def _evaluate_assertions(adapter: PlatformAdapter, assertions) -> Tuple[List[Dict], bool, List[str]]:
    results: List[Dict] = []
    ok = True
    failed_targets: List[str] = []
    for a in assertions:
        status = "passed"
        details = ""
        if a.type == "visible":
            visible = adapter.exists(a.target)
            if not visible:
                status = "failed"
                details = f"Target not visible: {a.target}"
                ok = False
                failed_targets.append(a.target)
        else:
            status = "failed"
            details = f"Unsupported assertion type: {a.type}"
            ok = False
            failed_targets.append(a.target)

        results.append({"id": a.id, "type": a.type, "target": a.target, "status": status, "details": details})
    return results, ok, failed_targets


def _signature_safe_details(details: Dict) -> Dict:
    safe = copy.deepcopy(details)
    args = safe.get("args")
    if isinstance(args, dict):
        # path includes run-id and varies run-to-run; exclude from determinism signature
        args.pop("path", None)
    return safe


def run_scenario(scenario, adapter: PlatformAdapter, device: str = "local", timeout_ms: int | None = None, artifact_base_dir: str | None = None) -> RunResult:
    started_at = utc_now_iso()
    run_id = make_id("run")
    platform = scenario.meta["platform"]
    step_results: List[StepResult] = []
    artifacts: List[str] = []
    top_error = None

    screenshot_dir: Path | None = None
    source_dir: Path | None = None
    if artifact_base_dir:
        screenshot_dir = Path(artifact_base_dir) / run_id / "screenshots"
        screenshot_dir.mkdir(parents=True, exist_ok=True)
        source_dir = Path(artifact_base_dir) / run_id / "sources"
        source_dir.mkdir(parents=True, exist_ok=True)

    try:
        for step in scenario.steps:
            start = time.perf_counter()
            try:
                step_args = dict(step.args)
                if step.command == "screenshot" and screenshot_dir is not None:
                    label = step_args.get("label", step.id)
                    step_args["path"] = str(screenshot_dir / f"{label}.png")
                if step.command == "source" and source_dir is not None:
                    label = step_args.get("label", step.id)
                    step_args["path"] = str(source_dir / f"{label}.xml")

                details = _run_step(adapter, step.command, step_args)
                duration = int((time.perf_counter() - start) * 1000)
                result = StepResult(
                    id=step.id,
                    command=step.command,
                    status=Status.OK,
                    duration_ms=duration,
                    details=details,
                )
                if step.command in {"screenshot", "source"}:
                    artifacts.append(details["args"].get("path", details["args"].get("file", "")))
                if timeout_ms and duration > timeout_ms:
                    result.status = Status.FAIL
                    result.error = make_error(
                        ErrorCode.ACTION_ERROR,
                        f"Step '{step.id}' exceeded timeout",
                        "Device/app was too slow for configured timeout",
                        "Increase --timeout or optimize target action",
                    )
            except Exception as exc:
                duration = int((time.perf_counter() - start) * 1000)
                result = StepResult(
                    id=step.id,
                    command=step.command,
                    status=Status.FAIL,
                    duration_ms=duration,
                    details={},
                    error=make_error(
                        ErrorCode.ACTION_ERROR,
                        f"Step '{step.id}' failed",
                        str(exc),
                        "Inspect step args and platform adapter support",
                    ),
                )
            step_results.append(result)

        assertions, asserts_ok, failed_targets = _evaluate_assertions(adapter, scenario.assertions)

        steps_ok = all(s.status == Status.OK for s in step_results)
        if not steps_ok:
            top_error = make_error(
                ErrorCode.ACTION_ERROR,
                "Run failed due to one or more action step failures",
                "At least one command execution step returned failure",
                "Inspect failed step error payloads in run.steps and retry",
            )
        if not asserts_ok:
            first_failed = failed_targets[0] if failed_targets else "unknown target"
            top_error = make_error(
                ErrorCode.ASSERTION_ERROR,
                "Run failed due to assertion failure",
                f"Expected UI target was not satisfied: {first_failed}",
                "Verify selector correctness and app state before assertion step",
            )

        all_pass = steps_ok and asserts_ok
        ended_at = utc_now_iso()
        signature_input = {
            "platform": platform,
            "steps": [
                {
                    "id": s.id,
                    "command": s.command,
                    "status": s.status.value,
                    "details": _signature_safe_details(s.details),
                }
                for s in step_results
            ],
            "assertions": assertions,
        }

        return RunResult(
            run_id=run_id,
            platform=platform,
            device=device,
            started_at=started_at,
            ended_at=ended_at,
            status=Status.OK if all_pass else Status.FAIL,
            steps=step_results,
            assertions=assertions,
            artifacts=artifacts,
            determinism_signature=signature_for(signature_input),
            seed=scenario.config.get("seed"),
            error=top_error,
        )
    finally:
        adapter.close()


def determinism_check(signatures: List[str]) -> float:
    if not signatures:
        return 0.0
    baseline = signatures[0]
    same = sum(1 for s in signatures if s == baseline)
    return round((same / len(signatures)) * 100, 2)

File: src/runner.ts
import fs from 'node:fs';
import path from 'node:path';
import { performance } from 'node:perf_hooks';

import { makeError } from './errors.js';
import type {
  Assertion,
  AssertionResult,
  PlatformAdapter,
  RunResult,
  Scenario,
  StepResult
} from './types.js';
import { makeId, signatureFor, utcNowIso } from './utils.js';

async function runStep(
  adapter: PlatformAdapter,
  command: Scenario['steps'][number]['command'],
  args: Record<string, unknown>
): Promise<Record<string, unknown>> {
  switch (command) {
    case 'tap':
      return adapter.tap(args);
    case 'navigate':
      return adapter.navigate(args);
    case 'act':
      return adapter.act(args);
    case 'screenshot':
      return adapter.screenshot(args);
    case 'wait':
      return adapter.wait(args);
    case 'source':
      return adapter.source(args);
  }
}

async function evaluateAssertions(
  adapter: PlatformAdapter,
  assertions: Assertion[]
): Promise<{ results: AssertionResult[]; ok: boolean; failedTargets: string[] }> {
  const results: AssertionResult[] = [];
  let ok = true;
  const failedTargets: string[] = [];

  for (const assertion of assertions) {
    let status: AssertionResult['status'] = 'passed';
    let details = '';

    if (assertion.type === 'visible') {
      const visible = await adapter.exists(assertion.target);
      if (!visible) {
        status = 'failed';
        details = `Target not visible: ${assertion.target}`;
        ok = false;
        failedTargets.push(assertion.target);
      }
    } else {
      status = 'failed';
      details = `Unsupported assertion type: ${assertion.type}`;
      ok = false;
      failedTargets.push(assertion.target);
    }

    results.push({
      id: assertion.id,
      type: assertion.type,
      target: assertion.target,
      status,
      details
    });
  }

  return { results, ok, failedTargets };
}

function signatureSafeDetails(details: Record<string, unknown>): Record<string, unknown> {
  const safe = structuredClone(details);
  const args = safe.args;
  if (args && typeof args === 'object' && !Array.isArray(args)) {
    delete (args as Record<string, unknown>).path;
  }
  return safe;
}

export async function runScenario(
  scenario: Scenario,
  adapter: PlatformAdapter,
  device = 'local',
  timeoutMs?: number,
  artifactBaseDir?: string
): Promise<RunResult> {
  const started_at = utcNowIso();
  const run_id = makeId('run');
  const platform = scenario.meta.platform;
  const stepResults: StepResult[] = [];
  const artifacts: string[] = [];
  let topError: RunResult['error'];

  let screenshotDir: string | undefined;
  let sourceDir: string | undefined;
  if (artifactBaseDir) {
    screenshotDir = path.join(artifactBaseDir, run_id, 'screenshots');
    sourceDir = path.join(artifactBaseDir, run_id, 'sources');
    fs.mkdirSync(screenshotDir, { recursive: true });
    fs.mkdirSync(sourceDir, { recursive: true });
  }

  try {
    for (const step of scenario.steps) {
      const started = performance.now();

      try {
        const stepArgs = { ...step.args };
        if (step.command === 'screenshot' && screenshotDir) {
          const label = String(stepArgs.label ?? step.id);
          stepArgs.path = path.join(screenshotDir, `${label}.png`);
        }
        if (step.command === 'source' && sourceDir) {
          const label = String(stepArgs.label ?? step.id);
          stepArgs.path = path.join(sourceDir, `${label}.xml`);
        }

        const details = await runStep(adapter, step.command, stepArgs);
        const durationMs = Math.round(performance.now() - started);
        const result: StepResult = {
          id: step.id,
          command: step.command,
          status: 'ok',
          duration_ms: durationMs,
          details
        };

        if (step.command === 'screenshot' || step.command === 'source') {
          const args = details.args;
          if (args && typeof args === 'object' && !Array.isArray(args)) {
            const artifactPath =
              typeof (args as Record<string, unknown>).path === 'string'
                ? ((args as Record<string, unknown>).path as string)
                : typeof (args as Record<string, unknown>).file === 'string'
                  ? ((args as Record<string, unknown>).file as string)
                  : '';
            if (artifactPath) {
              artifacts.push(artifactPath);
            }
          }
        }

        if (timeoutMs && durationMs > timeoutMs) {
          result.status = 'fail';
          result.error = makeError(
            'ACTION_ERROR',
            `Step '${step.id}' exceeded timeout`,
            'Device/app was too slow for configured timeout',
            'Increase --timeout or optimize target action'
          );
        }

        stepResults.push(result);
      } catch (error) {
        const durationMs = Math.round(performance.now() - started);
        stepResults.push({
          id: step.id,
          command: step.command,
          status: 'fail',
          duration_ms: durationMs,
          details: {},
          error: makeError(
            'ACTION_ERROR',
            `Step '${step.id}' failed`,
            error instanceof Error ? error.message : String(error),
            'Inspect step args and platform adapter support'
          )
        });
      }
    }

    const assertionEvaluation = await evaluateAssertions(adapter, scenario.assertions);
    const stepsOk = stepResults.every((step) => step.status === 'ok');

    if (!stepsOk) {
      topError = makeError(
        'ACTION_ERROR',
        'Run failed due to one or more action step failures',
        'At least one command execution step returned failure',
        'Inspect failed step error payloads in run.steps and retry'
      );
    }

    if (!assertionEvaluation.ok) {
      const firstFailedTarget = assertionEvaluation.failedTargets[0] ?? 'unknown target';
      topError = makeError(
        'ASSERTION_ERROR',
        'Run failed due to assertion failure',
        `Expected UI target was not satisfied: ${firstFailedTarget}`,
        'Verify selector correctness and app state before assertion step'
      );
    }

    const allPass = stepsOk && assertionEvaluation.ok;
    const ended_at = utcNowIso();
    const signatureInput = {
      platform,
      steps: stepResults.map((step) => ({
        id: step.id,
        command: step.command,
        status: step.status,
        details: signatureSafeDetails(step.details)
      })),
      assertions: assertionEvaluation.results
    };

    return {
      run_id,
      platform,
      device,
      started_at,
      ended_at,
      status: allPass ? 'ok' : 'fail',
      steps: stepResults,
      assertions: assertionEvaluation.results,
      artifacts,
      determinism_signature: signatureFor(signatureInput),
      seed: typeof scenario.config.seed === 'number' ? scenario.config.seed : undefined,
      error: topError
    };
  } finally {
    await adapter.close();
  }
}

export function determinismCheck(signatures: string[]): number {
  if (signatures.length === 0) {
    return 0;
  }

  const baseline = signatures[0];
  const same = signatures.filter((signature) => signature === baseline).length;
  return Math.round((same / signatures.length) * 10000) / 100;
}
